src/storage-area.cpp

```diff
@@ -0,0 +1,202 @@
+        file.write(content.get(),size);
+        file.close();
+        // set file permissions
+        // todo: permission debug info?
+        fs::file_status master_status = fs::status(master_path);
+        master_status.permissions(globals::file_permissions);
+
+        // create hard links
+        auto hardlink_to = [&](std::string groupby, std::string group) {
+            fs::path link = fs::path(storage_root)
+                    .append(groupby)
+                    .append(group)
+                    .append(uuid)
+                    .append(".DCM");
+            fs::create_directories(link);
+            fs::create_hard_link(master_path, link);
+            fs::permissions(link, globals::file_permissions);
+        };
+        // todo: integrate json settings to enable/disable individual hard links
+        // todo: replace placeholders
+        std::string DOB_placeholder;
+        std::string PID_placeholder;
+        std::string SD_placeholder;
+        hardlink_to("/by-dob/", DOB_placeholder);
+        hardlink_to("/by-patient-id/", PID_placeholder);
+        hardlink_to("/by-study-date/", SD_placeholder);
+        return OrthancPluginErrorCode_Success;
+    }
+    return OrthancPluginErrorCode_BadFileFormat;
+}
+
+OrthancPluginErrorCode StorageCreateCallback(const char *uuid,
+                                             const void *content,
+                                             int64_t size,
+                                             OrthancPluginContentType type) {
+    fs::path path;
+    switch (type) {
+        case OrthancPluginContentType_Dicom:
+            return WriteDicomFile(DicomFile(content, size), uuid);
+        case OrthancPluginContentType_DicomAsJson:
+        case OrthancPluginContentType_DicomUntilPixelData:
+        case _OrthancPluginContentType_INTERNAL:
+        case OrthancPluginContentType_Unknown:
+            path = GetPath(type,uuid);
+    }
+    std::fstream file(path, std::ios::out | std::ios::binary);
+    if (file.is_open()) {
+        file.write((const char *) content, size);
+        if (file.good()) {
+            file.close();
+            return OrthancPluginErrorCode_Success;
+        }
+        OrthancPluginLogWarning(globals::context, "StorageCreateCallback: but write out appears bad");
+    }
+    return OrthancPluginErrorCode_FileStorageCannotWrite;
+}
+
+OrthancPluginErrorCode StorageReadWholeCallback(OrthancPluginMemoryBuffer64 *target,
+                                                const char *uuid,
+                                                OrthancPluginContentType type) {
+    auto path = GetPath(type,uuid);
+    // todo: is the buffer ready? without this call..
+    OrthancPluginCreateMemoryBuffer64(globals::context, target, fs::file_size(path));
+    std::fstream file(path, std::ios::in | std::ios::binary);
+    if(file.is_open()){
+        file.read((char*)target->data, target->size);
```

Comment:
Hopefully it is just this easy. Not 100% sure what data buffer formats will be written to disk in the Create callback.. but this will read them as they were received [for writing] in Create. If testing reveals malformed json files or other problems, then we'll need to revise this along with the write procedure. Again, hopefully it is just this simple that we can write the buffer to disk and read it just as easy.